### ISSUE TITLE
HTTP2: Support bidirectional content

### ISSUE THREAD

**Issue by Tratcher** [dotnet/runtime#1511]
## Update

As part of 3.0 we essentially disabled bidirectional HttpContent, implementing only what was needed for gRPC support. We now need to, as part of 5.0/Future, consider how to enable general-purpose bidirectional support.

## Original issue
https://github.com/dotnet/corefx/pull/36884 (HTTP/2 100 Continue) also snuck in bidirectional support (https://github.com/dotnet/corefx/issues/36855) for HTTP/2. This works pretty well but does introduce an ambiguous scenario when the server completes the response before the client completes the request.

Server code: (Kestrel 3.0.0-preview7)
```c#
                var response = "hello, world";
                context.Response.ContentLength = response.Length;
                await context.Response.WriteAsync(response);

                await context.Features.Get<IHttpResponseCompletionFeature>().CompleteAsync();
                await Task.Delay(TimeSpan.FromSeconds(5));
                // Drain the request body
                await context.Request.Body.CopyToAsync(Stream.Null);
```
Yes this server code is a bit unusual, it was written this way to make the error visible from CopyToAsync. You'll get similar behavior if you just write Hello World and let the request unwind.

Client code:
```c#
            var client = new HttpClient();
            client.DefaultRequestVersion = new Version(2, 0);
            using var file = File.OpenRead(@"A200MbFile.txt");
            var content = new StreamContent(file);
            var response = await client.PostAsync("https://localhost:5001", content);
```
- In HTTP/1.x this works fine, PostAsync won't complete until the request stream has been completely uploaded. 
- In HTTP/2 it's ambiguous because when PostAsync completes you have a fully buffered response body but you have no idea what the state of the request is nor a way to be notified of its completion. Trying to perform any operations on that file can cause errors in the client code because it's still in use. If you close the file now the request will fail, but the error will only be visible on the server.
- HttpCompletionOptions do not have any affect on the request, they only affect the response.
- None of the built in HttpContent implementations provide any visibility on their progress. PushStreamContent from WebApi is the only HttpContent example I've seen that would surface this state to users.

Mitigations:
- Servers usually wait for the request body to end before ending the response body. However the client can't always rely on the server behaving a specific way, ending the response early is within spec.
- For >= 300 responses the HTTP/2 stream will be reset and the request stream abandon. These are the most common scenarios where the server would end the response without draining the request. This doesn't surface any errors directly to the caller unless they have a custom HttpContent, they have to infer the request body is done/abandon from the status code.

These mitigations make the issue occur infrequently enough that it will cause some very non-obvious issues when it it does happen (e.g. locked files, InvalidOperationExceptions, multi-threaded stream re-use, etc.).

Proposals:
A) At a minimum there needs to be guidance given that this might happen and how the user might handle it. I don't know what guidance could be given beyond implementing a custom HttpContent.
B) Add something to HttpContent to make progress more visible. E.g. something they could await or DisposeAsync that could wait for things to finish first. This wouldn't be very discoverable and would need to be accompanied by guidance.
C) Only enable bidirectional support when using HttpCompletionOptions.ResponseHeadersRead (or a new flag?). Most people don't use this flag so they wouldn't be impacted by the behavior change (e.g. PostAsync doesn't support that mode). I'm not sure how you implement this without B.

@wfurt @davidsh

**Comment by alrz** [comment]
Is this an ask for `PushStreamContent`? I was surprised that we don't have that type in core.

**Comment by Tratcher** [comment]
@alrz PushStreamContent is mostly unrelated. The ask here is for HttpClient to allow streaming the response data before the request data upload is finished. PushStreamContent might be convenient for providing the request content, but wouldn't affect the bidirectionality.

**Comment by alrz** [comment]
Currently we already can stream *from* HttpClient, to stream *to* HttpClient I think we need something like PushStreamContent. Should I file an issue for that?

**Comment by Tratcher** [comment]
@alrz yes please.

**Comment by alrz** [comment]
Files https://github.com/dotnet/runtime/issues/48386

**Comment by ryan-motive** [comment]
Has there been any movement on this issue? More generally, is there any way to wrestle HttpClient to work as a truly bi-directional HTTP/2 connection? We're talking to an HTTP/2 service and at a bit of a loss for how best to tackle this in C#. Are there any options besides building out a custom implementation using TcpClient?

**Comment by alexandrehtrb** [comment]
+1

> Are there any options besides building out a custom implementation using TcpClient?

I think WebSockets is an alternative...

**Comment by ryan-motive** [comment]
Yep, using Websockets currently and it's working well, but it feels like a placeholder until .NET gets proper support for HTTP/2 streaming.

**Comment by karelz** [comment]
@ryan-motive can you share some details about your scenario? Why do you need it? Is it server requirement you need to match?

**Comment by ryan-motive** [comment]
It's not really relevant why I need it--the spec was written in 2015 and yet this part of the spec is still not supported by the .NET networking library. This seems like a pretty significant missing piece.

**Comment by karelz** [comment]
@ryan-motive I respectfully disagree. There are things RFCs specify which are rarely used and not all stacks implement them. We believe this might be one of them in HTTP/2 RFC.
In general, we prioritize work based on customer demand -- when we know about user scenarios requiring the features and/or when there are lots of +1s. If you want to help us prioritize the work in upcoming releases, then sharing details about your scenario will help us with that.

I also suggest everyone who wishes to have this feature to upvote top post. Thanks!

**Comment by ryan-motive** [comment]
Then at a minimum make this the official policy and include it in the documentation so that we don't spend days digging through Github forums trying to figure out why the feature is not there.

**Comment by stephentoub** [comment]
> More generally, is there any way to wrestle HttpClient to work as a truly bi-directional HTTP/2 connection?

I'm a little confused. This is already possible with HTTP/2, has been since .NET Core 3.0; that's how gRPC works.  What specifically isn't working?  The core of this issue was addressed as described in https://github.com/dotnet/runtime/issues/1511#issuecomment-509864597.

**Comment by ryan-motive** [comment]
I don't see how this solves the problem but maybe I'm not understanding that comment? Basically I just need a stream for writing and a stream for reading a duplex connection. I wasn't able to find a solution and this thread suggested there wasn't one. I do recall reading your comment when I was trying to solve this, but frankly I didn't really understand it 😆 

You're correct though--the use cases here are gRPC and there are a number of AWS services that use HTTP/2 streaming. Right now it's not possible to use C# to access those so we're having to go through a C++ wrapper which adds complexity.

**Comment by davidfowl** [comment]
I wanted to reply with what @stephentoub said but I was waiting for a code sample to show what the problem was. I wrote this a long time ago, what am I missing https://github.com/davidfowl/StreamingSample?

**Comment by stephentoub** [comment]
Right, what David just said. A custom HttpContent-derived type set on the request supports copying concurrently with the HttpContent on the response stream. So you can author an HttpContent that's the send side that you write to for the duration of the operation, and consume from the response HttpContent for the duration of the operation.  The comments in this thread about it not working were from before we allowed it to work with HTTP/2.

**Comment by ryan-motive** [comment]
Wow, I wish I'd seen this! It's long in the past now, but this is helpful to know for the future.

**Comment by karelz** [comment]
@stephentoub what is left to track on this issue then?

**Comment by stephentoub** [comment]
> @stephentoub what is left to track on this issue then?

https://github.com/dotnet/runtime/issues/1511#issuecomment-538574216
"Notes re remaining work: We added an internal "AllowDuplex" property on HttpContent. We should consider making this public."

**Comment by karelz** [comment]
Thanks @stephentoub! Top post updated to clarify for future.
